src/registry.c

```diff
@@ -0,0 +1,1196 @@
+    va_end(args);
+
+    /* This shouldn't really happen */
+    if (rc < 0) {
+        log_err(ctx, "+++ out of cheese error. redo from start +++\n");
+        slen = 0;
+        memset(buf, 0, sizeof(buf));
+        return;
+    }
+
+    slen += rc;
+    if (slen >= (int)sizeof(buf)) {
+        /* truncated, let's flush this */
+        buf[sizeof(buf) - 1] = '\n';
+        slen = sizeof(buf);
+    }
+
+    /* We're assuming here that the last character is \n. */
+    if (buf[slen - 1] == '\n') {
+        log_err(ctx, "%s", buf);
+        memset(buf, 0, sizeof(buf));
+        slen = 0;
+    }
+}
+
+static bool
+validate(struct rxkb_context *ctx, xmlDoc *doc)
+{
+    bool success = false;
+    xmlValidCtxt *dtdvalid = NULL;
+    xmlDtd *dtd = NULL;
+    xmlParserInputBufferPtr buf = NULL;
+    /* This is a modified version of the xkeyboard-config xkb.dtd, we
+     * allow for more either of modelList, layoutList or optionList to be
+     * missing.
+     */
+    const char dtdstr[] =
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+        "<!ELEMENT xkbConfigRegistry (modelList?, layoutList?, optionList?)>\n"
+        "<!ATTLIST xkbConfigRegistry version CDATA \"1.1\">\n"
+        "<!ELEMENT modelList (model*)>\n"
+        "<!ELEMENT model (configItem)>\n"
+        "<!ELEMENT layoutList (layout*)>\n"
+        "<!ELEMENT layout (configItem,  variantList?)>\n"
+        "<!ELEMENT optionList (group*)>\n"
+        "<!ELEMENT variantList (variant*)>\n"
+        "<!ELEMENT variant (configItem)>\n"
+        "<!ELEMENT group (configItem, option*)>\n"
+        "<!ATTLIST group allowMultipleSelection (true|false) \"false\">\n"
+        "<!ELEMENT option (configItem)>\n"
+        "<!ELEMENT configItem (name, shortDescription?, description?, vendor?, countryList?, languageList?, hwList?)>\n"
+        "<!ATTLIST configItem popularity (standard|exotic) \"standard\">\n"
+        "<!ELEMENT name (#PCDATA)>\n"
+        "<!ELEMENT shortDescription (#PCDATA)>\n"
+        "<!ELEMENT description (#PCDATA)>\n"
+        "<!ELEMENT vendor (#PCDATA)>\n"
+        "<!ELEMENT countryList (iso3166Id+)>\n"
+        "<!ELEMENT iso3166Id (#PCDATA)>\n"
+        "<!ELEMENT languageList (iso639Id+)>\n"
+        "<!ELEMENT iso639Id (#PCDATA)>\n"
+        "<!ELEMENT hwList (hwId+)>\n"
+        "<!ELEMENT hwId (#PCDATA)>\n";
+
+    /* Note: do not use xmlParserInputBufferCreateStatic, it generates random
+     * DTD validity errors for unknown reasons */
+    buf = xmlParserInputBufferCreateMem(dtdstr, sizeof(dtdstr),
+                                        XML_CHAR_ENCODING_UTF8);
+    if (!buf)
+        return false;
+
+    dtd = xmlIOParseDTD(NULL, buf, XML_CHAR_ENCODING_UTF8);
+    if (!dtd) {
+        log_err(ctx, "Failed to load DTD\n");
+        return false;
+    }
+
+    dtdvalid = xmlNewValidCtxt();
+    if (!xmlValidateDtd(dtdvalid, doc, dtd)) {
+        log_err(ctx, "XML error: failed to validate document\n");
```

Comment:
Can we include the path in this message?

Reply:
fixed by moving the error msg out into the caller